bin/dart

```diff
@@ -0,0 +1,186 @@
+  return 0
+}
+
+function upgrade_flutter () {
+  mkdir -p "$FLUTTER_ROOT/bin/cache"
+
+  # This function is executed with a redirect that pipes the source of
+  # this script into file descriptor 3.
+  #
+  # To ensure that we don't simultaneously update Dart in multiple
+  # parallel instances, we try to obtain an exclusive lock on this
+  # file descriptor (and thus this script's source file) while we are
+  # updating Dart and compiling the script. To do this, we try to use
+  # the command line program "flock", which is available on many
+  # Unix-like platforms, in particular on most Linux distributions.
+  # You give it a file descriptor, and it locks the corresponding
+  # file, having inherited the file descriptor from the shell.
+  #
+  # Complicating matters, there are two major scenarios where this
+  # will not work.
+  #
+  # The first is if the platform doesn't have "flock", for example on Mac.
+  # There is not a direct equivalent, so on platforms that don't have flock,
+  # we fall back to using a lockfile and spinlock with "shlock".  This
+  # doesn't work as well over NFS as it relies on PIDs. Any platform
+  # without either of these tools has no locking at all. To determine if we
+  # have "flock" or "shlock" available, we abuse the "hash" shell built-in.
+  #
+  # The second complication is NFS. On NFS, to obtain an exclusive
+  # lock you need a file descriptor that is open for writing, because
+  # NFS implements exclusive locks by writing, or some such. Thus, we
+  # ignore errors from flock. We do so by using the '|| true' trick,
+  # since we are running in a 'set -e' environment wherein all errors
+  # are fatal, and by redirecting all output to /dev/null, since
+  # users will typically not care about errors from flock and are
+  # more likely to be confused by them than helped.
+  #
+  # For "flock", the lock is released when the file descriptor goes out of
+  # scope,  i.e. when this function returns.  The lock is released via
+  # a trap when using "shlock".
+  if hash flock 2>/dev/null; then
+    flock 3 2>/dev/null || true
+  elif hash shlock 2>/dev/null; then
+    FLUTTER_UPGRADE_LOCK="$FLUTTER_ROOT/bin/cache/.upgrade_lock"
+    while ! shlock -f "$FLUTTER_UPGRADE_LOCK" -p $$ ; do sleep .1 ; done
+    trap _rmlock EXIT
+  fi
+
+  local revision="$(cd "$FLUTTER_ROOT"; git rev-parse HEAD)"
+
+  # Invalidate cache if:
+  #  * SNAPSHOT_PATH is not a file, or
+  #  * STAMP_PATH is not a file with nonzero size, or
+  #  * Contents of STAMP_PATH is not our local git HEAD revision, or
+  #  * pubspec.yaml last modified after pubspec.lock
+  if [[ ! -f "$SNAPSHOT_PATH" || ! -s "$STAMP_PATH" || "$(cat "$STAMP_PATH")" != "$revision" || "$FLUTTER_TOOLS_DIR/pubspec.yaml" -nt "$FLUTTER_TOOLS_DIR/pubspec.lock" ]]; then
+    rm -f "$FLUTTER_ROOT/version"
+    touch "$FLUTTER_ROOT/bin/cache/.dartignore"
+    "$FLUTTER_ROOT/bin/internal/update_dart_sdk.sh"
+    VERBOSITY="--verbosity=error"
+
+    echo Building flutter tool...
```

Comment:
I'm not sure how to handle this, but I expect users are going to be confused why invoking dart is re-building their flutter tool.

If we exit early here, without updating `$STAMP_PATH`, the next time the user invokes the flutter tool it will rebuild. Although it means disrupting the user's flow twice, I think at least that flow is more expected. WDYT?

Reply:
I don't think we should rebuild their `flutter` tool unless they're actually using the `flutter` tool.

Why do you say this disrupts them twice?  Isn't it just the once when they try to invoke the `flutter` tool (as it is today)?  I mean, sure, it invalidates the cache earlier this way, but I don't think the user will notice that until they run flutter, and the `flutter` tool is going to rebuild at most once for each cache invalidation anyhow. At least this will have updated the dart SDK earlier for them.

Plus, if they don't need to run the flutter tool, then we're wasting their time/energy/battery rebuilding it.

Reply:
keep in mind that as contributors to flutter you will hide this code path approximately one million times more often than someone who upgrades stable -> stable.

Let's keep it simple and reduce conditional logic to the minimum.

Reply:
TYPO: hide -> hit